### PR TITLE
Navbar Improvements: Updated Heuristics

### PR DESCRIPTION
**Pull request by sloansparger**
### Description

Implements new logic for the navbar being open/closed for a page and auto-closing in certain conditions.

This is the core change to complete https://github.com/metabase/metabase/issues/36370

### Technical description

The main part of this change is having the navbar state subscribe to react-router-redux `@@router/LOCATION_CHANGE`  event to decide if we should keep the navbar open/closed on route changes. A special case is accommodated for home page redirects where we want to treat a url we would typically auto-close for as if it were the normal home page. This is accomplished by adding a flag to the location state on the replace event which is used to return the prev state / ignore the event.

### How to verify

- Log into metabase, should see navbar open
- Visit a dashboard, should see navbar auto-close
- Click the MB logo, should stay closed
- Go back a page
- Manually open the navbar
- Click the MB logo, should stay open
- Go to a collection, should stay open
- Go to a dashboard within a collection, should auto-close
- Using the breadcrumbs at the top of the dashboard, go back to the collection, should stay closed
- Manually open the navbar
- Create a new question or sql query, should auto close

### Demo

https://github.com/metabase/metabase/assets/7104357/fc7c30c2-abe5-4504-b150-1ba083e23d71
(demo hangs at the end due to a perf issue unrelated to this change)

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

### CODE REVIEW

File: e2e/test/scenarios/navbar.cy.spec.js [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+      cy.visit("/");
+      cy.findByTestId("main-navbar-root").should("be.visible");
+    });
+
+    it("state should preserve when clicking the mb logo", () => {
+      cy.visit("/collection/root");
+      cy.findByTestId("main-navbar-root").should("be.visible");
+      cy.findByTestId("main-logo-link").click();
+      cy.findByTestId("main-navbar-root").should("be.visible");
+      visitDashboard(8);
```
Comment (sloansparger) [inline e2e/test/scenarios/navbar.cy.spec.js:26]: anyone know where I can import constants for things like this. i make a few magical references in this file that i'd like to fix...

File: frontend/src/metabase/nav/containers/MainNavbar/MainNavbar.styled.tsx [hunk 2]
```diff
@@ -140,7 +140,7 @@ export const LoadingTitle = styled.h2`
 `;
 
 export const PaddedSidebarLink = styled(SidebarLink)`
-  padding-left: ${space(2)};
+  padding-left: 12px;
```
Comment (iethree) [inline frontend/src/metabase/nav/containers/MainNavbar/MainNavbar.styled.tsx:143]: nit: I think we prefer 0.75rem

File: e2e/support/helpers/e2e-ui-elements-helpers.js [hunk 3]
```diff
@@ -34,7 +34,7 @@ export function leftSidebar() {
 }
 
 export function navigationSidebar() {
-  return cy.get("#root aside").first();
+  return cy.findByTestId("main-navbar-root");
```
Comment (iethree) [inline e2e/support/helpers/e2e-ui-elements-helpers.js:37]: ❤️

**Review by iethree — APPROVED** [review]
I love everything about this 🎉 


200w-1

(but I still found a few little nits that you could address)

**Comment by iethree** [comment]
But really, I just love how you consolidated all the random bits of sidebar logic so cleanly. 💖

File: e2e/test/scenarios/navbar.cy.spec.js
[the remark this answers is not in the document]
Comment (sloansparger) [reply e2e/test/scenarios/navbar.cy.spec.js:1]: I think I don't actually need this. Good callout.

File: frontend/src/metabase/redux/app.ts [hunk 4]
```diff
@@ -12,6 +12,19 @@ import {
 
 import type { Dispatch } from "metabase-types/store";
 
+interface LocationChangeAction {
+  type: string; // "@@router/LOCATION_CHANGE"
```
[the remark this answers is not in the document]
Comment (sloansparger) [reply frontend/src/metabase/redux/app.ts:16]: IT YELLS AT ME 😢 that's what I wanted to do, but I think there's some prerequisite typing we'd need to do elsewhere in our redux setup to get it happy again.

**Comment by luizarakaki** [comment]
Nice work! Everything working as specced

**Pull request merged**